tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp

```diff
@@ -0,0 +1,142 @@
+#define LINALG_TPLIMPLEMENTATIONS_INCLUDE_EXPERIMENTAL___P1673_BITS_KOKKOSKERNELS_BLAS2_MATRIX_RANK_1_UPDATE_HPP_
+
+#include <complex>
+
+namespace KokkosKernelsSTD {
+
+namespace Impl {
+
+// manages parallel execution of independent action
+// called like action(i, j) for each matrix element A(i, j)
+template <typename ExecSpace, typename MatrixType>
+class ParallelMatrixVisitor {
+public:
+  KOKKOS_INLINE_FUNCTION ParallelMatrixVisitor(ExecSpace &&exec_in, MatrixType &A_in):
+    exec(exec_in), A(A_in), ext0(A.extent(0)), ext1(A.extent(1))
+  {}
+
+  template <typename ActionType>
+  void for_each_triangle_matrix_element(std::experimental::linalg::upper_triangle_t t, ActionType action) {
+    Kokkos::parallel_for(Kokkos::RangePolicy(exec, 0, ext1),
+      KOKKOS_LAMBDA(const auto j) {
+        using idx_type = std::remove_const_t<decltype(j)>;
+        for (idx_type i = 0; i <= j; ++i) {
+          action(i, j);
+        }
+      });
+    exec.fence();
+  }
+
+  template <typename ActionType>
+  void for_each_triangle_matrix_element(std::experimental::linalg::lower_triangle_t t, ActionType action) {
+    for_each_triangle_matrix_element(std::experimental::linalg::upper_triangle,
+        [action](const auto i, const auto j) {
+          action(j, i);
+      });
+  }
+
+private:
+
+  template <typename T>
+  KOKKOS_INLINE_FUNCTION
+  static bool is_int_mul_safe(T a, T b) {
+    static constexpr auto max_val = std::numeric_limits<decltype(a * b)>::max();
+    return a <= max_val / b;
+  }
+
+private:
+  const ExecSpace &exec;
+  MatrixType &A;
+  const size_t ext0;
+  const size_t ext1;
+};
+
+} // namespace Impl
+
+namespace Impl {
+
+// Note: phrase it simply and the same as in specification ("has unique layout")
+template <typename Layout,
+          std::experimental::extents<>::size_type numRows,
+          std::experimental::extents<>::size_type numCols>
+inline constexpr bool is_unique_layout_v = Layout::template mapping<
+    std::experimental::extents<numRows, numCols> >::is_always_unique();
+
+template <typename Layout>
+struct is_layout_blas_packed: public std::false_type {};
+
+template <typename Triangle, typename StorageOrder>
+struct is_layout_blas_packed<
+  std::experimental::linalg::layout_blas_packed<Triangle, StorageOrder>>:
+    public std::true_type {};
+
+template <typename Layout>
+inline constexpr bool is_layout_blas_packed_v = is_layout_blas_packed<Layout>::value;
+
+// Note: will only signal failure for layout_blas_packed with diferent triangle
+template <typename Layout, typename Triangle>
+struct triangle_layout_mismatch: public std::false_type {};
+
+template <typename StorageOrder, typename Triangle1, typename Triangle2>
+struct triangle_layout_mismatch<
+  std::experimental::linalg::layout_blas_packed<Triangle1, StorageOrder>,
+  Triangle2>
+{
+  static constexpr bool value = !std::is_same_v<Triangle1, Triangle2>;
```

Comment:
Would you consider instead defining `triangle_layout_match`?  It's harder for readers to do the double negative in their heads.

Reply:
Fixed with https://github.com/kokkos/stdBLAS/pull/167/commits/878fa923c8683602c949548fb6e5c9f666ecf622